### PR TITLE
Types and examples for scenes and wizards

### LINKED ISSUE #894: Type defenition of ctx.scene.state
**Is your feature request related to a problem? Please describe.**
I want to use state of scene to save data, but because it originally typed as object typescript 
does not allow this.

**Describe the solution you'd like**
It would be great to be able to set the type of state. For example pass an iterface to Telegraf or Stage

### LINKED ISSUE #1165: Wrong typings in 3.38.0
You have wrong typings in the latest release, js files have methods that are not declared in d.ts. For example, there is setChatStickerSet method in telegram.js and no declaration for it in telegram.d.ts

### LINKED ISSUE #1172: Typescript building error
Code 
```
import { Telegraf, Context, Stage, BaseScene, Middleware } from "telegraf"
import { MiddlewareFn } from "telegraf/typings/composer"
import { TelegrafContext } from "telegraf/typings/context"
import { BOT_TOKEN } from "./consts"

export const StartBot = async() => {
  const greeter = new BaseScene("greeter")
  greeter.enter((ctx) => {
    console.log("enter scene")
    ctx.reply("Hi")
  })

  const stage = new Stage([greeter])
  stage.register(greeter)

  const _bot = new Telegraf(BOT_TOKEN)
  _bot.use(stage.middleware())
  _bot.start((ctx) => {
    // ctx.scene.enter("greeter")
    // ctx.reply("Start input your class name and select from autocomplite")
  })
  await _bot.launch()
}
```
When trying to build throwing errors:
```
src/bot.ts(17,12): error TS2345: Argument of type 'MiddlewareFn<SceneContextMessageUpdate>' is not assignable to parameter of type 'Middleware<TelegrafContext>'.
  Type 'MiddlewareFn<SceneContextMessageUpdate>' is not assignable to type 'MiddlewareFn<TelegrafContext>'.
    Property 'scene' is missing in type 'TelegrafContext' but required in type 'SceneContextMessageUpdate'.
src/bot.ts(19,9): error TS2339: Property 'scene' does not exist on type 'TelegrafContext'.
```

### PR DESCRIPTION
**Pull request by KnorpelSenf**
# Description

The types of Stages, Scenes, and Wizards were mostly wrong. The examples did not typecheck.

This PR applies consistent types to this part of the library. The example bots are migrated to TS, they build now with minimal changes.

While the types should be correct everywhere now, this comes with the downside of complexity. Scenes rely on the session. As a result, everything is bloated up to having two type variables: one for the session data and one for the context object, as both types can be adjusted. This trickles down far.

Note that the type variables follow the standard convention of single character names rather than the special `telegraf` version of prepending a `T`. We can change this back if desired, I simply found it even less readable than the current version to make all names longer.

Fixes #1172 
Fixes #1165
Fixes #894

Helps with
- #705
- #1074
- #1139

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update (need an example bot that shows how to customize context and session data)

# How Has This Been Tested?

```bash
npm t
npm run build
npm run lint
```

# Checklist:

- [X] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [X] My changes generate no new warnings
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by wojpawlik — COMMENTED** [review]
I don't see anything obviously very wrong, but I understand pretty much nothing here...

File: docs/examples/scenes-bot.ts
Comment (KnorpelSenf) [reply docs/examples/scenes-bot.ts:28]: It is necessary to extend the bot context in one way or another if scenes should be used.

The default `SceneContext` interface can be used (or `extend`ed) to automatically get the correct types for `ctx.scene` on the context object. If it is omitted, only the regular `Context` properties are available, and hence
```ts
bot.command('greeter', (ctx) => ctx.scene.enter('greeter'))
```
would fail.

**Comment by KnorpelSenf** [comment]
> but I understand pretty much nothing here

Then check out what is going on if #1195 … no but seriously, do you think that I should write more comments?

**Comment by wojpawlik** [comment]
I don't know, better ask someone familiar with Scenes/Stages/Wizards

**Comment by KnorpelSenf** [comment]
> I don't know, better ask someone familiar with Scenes/Stages/Wizards

Who would that be?

I also had no idea about these things before I started working on this PR, so I'd appreciate it if someone with more experience than me could have a look.

**Comment by KnorpelSenf** [comment]
@dotcypress `git blame` told me that you an expert on this domain—would you mind reviewing this PR?

File: src/scenes/context.ts
Comment (wojpawlik) [inline src/scenes/context.ts:46]: 3 `this.ctx as any`s in this file have to go

**Review by wojpawlik — CHANGES_REQUESTED** [review]

File: src/scenes/context.ts
Comment (KnorpelSenf) [reply src/scenes/context.ts:46]: There is only three. How are you going to construct an object that has exactly one property which can be parametrized? I don't think TS allows us to do this, but I'm happy if you can prove me wrong

Comment (wojpawlik) [reply src/scenes/context.ts:46]:
```ts
C extends Context & { session?: { __scenes: something } }
```

Comment (KnorpelSenf) [reply src/scenes/context.ts:46]: That is wrong. You have to parametrize `session` with the `sessionName` argument of the constructor. Session data is not always stored in the same property on the context object.

Comment (KnorpelSenf) [reply src/scenes/context.ts:46]: We would maybe need some magic with string literal types as type variable for the property name, mapped types to dynamically construct the object, `keyof` things to make sure everything matches, … and there is zero benefit to the end user as everything is typed already.

If this was possible, we would be able to infer automatically generated context objects from the user in many cases.

Comment (wojpawlik) [reply src/scenes/context.ts:46]: I got read of that misfeature to simplify typings.

Comment (KnorpelSenf) [reply src/scenes/context.ts:46]: Well have a go

Comment (KnorpelSenf) [reply src/scenes/context.ts:46]: I mean, everything. Holy cow. How was I not aware of #1131?

I'll have to look into removing a large part of the type parameters everywhere (not just in this module). Takes some time.

Comment (KnorpelSenf) [reply src/scenes/context.ts]: Turns out I could not remove the parameters as I wanted to. Either way, I also removed the option to specify `sessionName` for scenes and wizards, hence removing the casts to `any`.

**Comment by KnorpelSenf** [comment]
@IMalyugin @wwwat @foodornt you all seem to be interested in having TypeScript types for Stages/Scenes.

The types are fixed now. However, I have never used that feature of `telegraf`, so I have quite little experience with scenes etc. Can you please pull in the code from this branch and verify that
- the types work as expected,
- the PR does not alter the behaviour of the feature, and
- you can customise the types with your own session data and context objects in order to adjust them to your needs?

This would be very helpful. We cannot merge this without other people's approval :)

Regarding the third point: I will write some examples about how to that. This is necessary anyway, but please don't hesitate to validate the first two things already!

**Comment by KnorpelSenf** [comment]
Some type variables are redundant. It has to be possible to infer certain variables across `session`, `scene.session`, and `session.__scenes`. It is annoying to type the names multiple times, and also it does not make sense to specify them differently in different positions.

This needs to be figured out before a merge.

**Comment by KnorpelSenf** [comment]
I added a lot of example bots. They illustrate every possible way to use the type variables. It's mostly redundant stuff, so it should be enough to check out the most basic one and the most advanced one for reviewing this PR. There is a lot of nesting of the same objects here. This makes it very convenient to use, so I don't think we should change it, but it also makes the types a bit annoying internally. They are pretty fine on the surface, though.

**Comment by KnorpelSenf** [comment]
I figured out https://github.com/telegraf/telegraf/pull/1194#issuecomment-741228992.

The problem is that `ctx.scene.session` is an alias for `ctx.session.__scenes`. It is in fact possible to add type annotations to `ctx.scene: SceneContextScene` that infer this correctly, however, as soon as we extend scenes and add wizards, this logic breaks. It is not possible to keep up this alias without larger restructurings of the module, which I'd like to avoid.

As a consequence, whenever the scene session under `ctx.scene.session` is augmented, this type has to be specified twice, once in a place that is responsible for `ctx.scene.session` and once for `ctx.session.scenes`. I added a large number of examples that illustrate this.

It is difficult to explain to new users why type inference works correctly for scenes but not for wizards in this place. Hence, it makes sense to simply advise to consequently specify both type variables for `SceneContext` and `WizardContext` whenever the scene/wizard-specific session is extended. Nothing can go wrong here, it may just be redundant in very special cases that are too rare to care about.

**Comment by wojpawlik** [comment]
Please review and test #1224, after it's merged you can merge `develop` here to fix the conflict and `// Enable graceful stop` to all the new examples. I'll review this in the meantime.

**Comment by wojpawlik** [comment]
I'd still be happy to get review from someone more experienced with the stuff before I approve.

File: docs/examples/scenes/scenes-bot.ts
Comment (wojpawlik) [inline docs/examples/scenes/scenes-bot.ts:47]: Duplicated
```suggestion
```

**Review by wojpawlik — CHANGES_REQUESTED** [review]

**Comment by KnorpelSenf** [comment]
For reviewing this, it helps a lot to install `ts-node` and run the example bots. You can
```
bot.use(async (c,n) => { console.log(c.session); await next(); console.log(c.session); })
```
to inspect that the scenes actually do their job.

The types are easy to check by verifying that the `scenes-bot-custom-context-and-session-and-scene-and-scene-session.ts` compiles, which it does.

**Review by wojpawlik — APPROVED** [review]
It'd be great if you sorted these out:
```
 /home/runner/work/telegraf/telegraf/src/scenes/context.ts
Warning:   101:10  warning  Unexpected nullable object value in conditional. An explicit null check is required  @typescript-eslint/strict-boolean-expressions
Warning:   120:10  warning  Unexpected nullable object value in conditional. An explicit null check is required  @typescript-eslint/strict-boolean-expressions

/home/runner/work/telegraf/telegraf/src/stage.ts
Warning:   27:12  warning  Unexpected string value in conditional. An explicit empty string check is required  @typescript-eslint/strict-boolean-expressions
```

Other than that, I don't see anything wrong here, but to be honest I understand almost nothing, so I just trust TypeScript (and you).

**Comment by KnorpelSenf** [comment]
If no one else shows up and pops a question until the end of tomorrow, I'll merge.

I have tested example bots for both scenes and wizards thoroughly and they worked without problems. Writing the other example bots made me confident that the types are correct.

Even if we missed something here, the existing types will be an order of magnitude better than the current state, so the rest is fixing bugs (and I can do that).

**Pull request merged**
